File: adapters/maze-path/src/main/java/sk/slsp/workshop/mazesolver/adapters/mazepath/adapter/MazePathInMemoryAdapter.java
package sk.slsp.workshop.mazesolver.adapters.mazepath.adapter;

import sk.slsp.workshop.mazesolver.Coordinates;
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.AddPlaceToStackPort;
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.DumpStackPort;
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.GetPlaceFromStackPort;
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.RemovePlaceFromStackPort;

import java.util.ArrayList;
import java.util.List;

public class MazePathInMemoryAdapter implements AddPlaceToStackPort, DumpStackPort, GetPlaceFromStackPort, RemovePlaceFromStackPort {
	
	private final List<Coordinates> path = new ArrayList<>();
	
	
	@Override
	public void addPlaceToStack(Coordinates coordinates) {
		path.add(coordinates);
	}
	
	
	@Override
	public List<Coordinates> dump() {
		return path;
	}
	
	
	@Override
	public Coordinates getPlaceFromStack() {
		return path.get(path.size() - 1);
	}
	
	
	@Override
	public Coordinates removePlaceFromStack() {
		return path.remove(path.size() - 1);
	}
	
}

File: adapters/maze-path/src/main/kotlin/sk/slsp/workshop/mazesolver/adapters/mazepath/adapter/MazePathInMemoryAdapter.kt
package sk.slsp.workshop.mazesolver.adapters.mazepath.adapter

import sk.slsp.workshop.mazesolver.Coordinates
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.AddPlaceToStackPort
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.DumpStackPort
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.GetPlaceFromStackPort
import sk.slsp.workshop.mazesolver.adapters.mazepath.port.RemovePlaceFromStackPort

class MazePathInMemoryAdapter:
	AddPlaceToStackPort,
	GetPlaceFromStackPort,
	RemovePlaceFromStackPort,
	DumpStackPort
{

	private val path: MutableList<Coordinates> = mutableListOf()

	override fun addPlaceToStack(coordinates: Coordinates) {
		path.add(coordinates)
	}

	override fun getPlaceFromStack(): Coordinates {
		return path.last()
	}

	override fun removePlaceFromStack(): Coordinates {
		return path.removeLast()
	}

	override fun dump(): List<Coordinates> {
		return path
	}

}
